### ISSUE TITLE
XQuartz won't start: hangs indefinitely or returns exit code 127

### ISSUE THREAD

**Issue by rmanzoni** [XQuartz/XQuartz#120]
Hi,

I'm having troubles with XQuartz recently but I cannot associate the inception of the problems to a well defined event.
For the record, I'm on Big Sur 11.2.3, but it's not obvious whether this has anything to do with the problem here.

The thing is that I cannot start XQuartz (or xterm or any other x program)
1. clicking on the XQuartz app --> exit code 127 
```
Mar 12 17:12:39 MacBook-Pro com.apple.xpc.launchd[1] (application.org.xquartz.X11.216728142.216728150[39551]): Service exited with abnormal code: 127
```
2. from terminal with `xterm`, `xquartz` etc --> hangs indefintely
3. whenever `/Applications/Utilities/XQuartz.app/Contents/MacOS/X11.sh` is executed, I get this error --> `Mar 12 16:42:52 MacBook-Pro org.macosforge.xquartz.startx[21311]: -: line 1: exec: complete: not found`, which points to `case $(basename "${SHELL}") in bash)          exec -l "${SHELL}" --login -c 'exec "${@}"' - "${@}" ;;`

The only way it starts is when `/Applications/Utilities/XQuartz.app/Contents/MacOS/X11.bin` is invoked explicitly.

The problem stays there regardless of whether I use 2.7.11 or 2.8.0_rc*.
`$DISPLAY` returns `/private/tmp/com.apple.launchd.c9WveOcJxH/org.xquartz:0` so it I guess it's fine.

I've read ostensibly all threads I could find and those that get close are of more or less variation of this
https://superuser.com/questions/699838/xquartz-will-not-start-on-mac-10-9-maverick-infinite-failure-loop
which points to possible problems in my login/profile/rc files.

So I've tried two approaches, one is to create a new virgin user on my mac and use XQuartz from there --> it worked fine, the other is to go use my normal user and start a new bash shell with `bash --norc --noprofile` such to circumvent shell initialisation messups --> this did not work.

In summary, from the test I've conducted I cannot tell whether the problem is due to my login/profile/rc files (and in that case I would have troubles figuring out where, I've sifted them already) or something else.

Thanks!

**Comment by jeremyhu** [comment]
Given that this is a configuration problem, it's best to ask for help on the mailing list.

That being said, you're best off moving aside any init scripts (~/.zprofile, ~/.zshrc, ~/.bash_profile, ~/.profile, ~/.bashrc, ~/.xinitrc, ~/.xinitrc.d, ~/.x11-run, etc) and then re-adding them until you find the one that is the problem.

**Comment by hraftery** [comment]
Lost a few hours on this. Turns out it's quite simple:

XQuartz no longer starts if your default shell is not `zsh`! After lots of wild-goose chases and looooong login/logout cycles to debug, and painful uninstall/re-install iterations with different package managers, I noticed my virgin account worked but started `zsh` by default. My main account has `bash` (a newer version, installed with Homebrew) as default.

So I simply issued `chsh -s /bin/zsh` and everything started working. I'm not sure what the scope of the setting is required to fix it, and I suspect a simple change to the XQuartz launcher would fix it at the heart. But I've got to move on with my life now!

I assume this is something to do with the default shell change in macOS and recent XQuartz releases trying to adapt. But man, it's insidious.

**Comment by jeremyhu** [comment]
No, that shouldn't be the case.  The startup script works fine with bash as well.  I suspect a bug in your bash shell init scripts (eg: ~/.bash_profile or similar).  That's usually the cause of bugs like this...

**Comment by hraftery** [comment]
Oh that makes sense. I did look there first, but I will look deeper. Must be something subtle.

**Comment by hraftery** [comment]
Argh, you're quite right @jeremyhu . I had a vestigial `set rmstar` in `.bash_profile` from waaaaaaaay back in my `tcsh` days. That's all! It has been sitting there unnoticed for a long time, and I was led far and wide in search of other issues.

I guess if someone has a similar issue, note the first message in this thread and try executing:

`/Applications/Utilities/XQuartz.app/Contents/MacOS/X11.sh`

Note that the `exec` statement will execute your shell's login scripts and any error you see might have a misleading line number and actually originate from within those scripts - even if you've never seen those scripts cause issues anywhere else!
